File: 백준/Gold/16398. 행성 연결/행성 연결.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] parents;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        List<int[]> li = new ArrayList<>();
        for (int i = 0 ; i<N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j = 0; j<N; j++){
                int n = Integer.parseInt(st.nextToken());
                li.add(new int[]{n, i, j});
            }
        }

        parents = new int[N];
        for(int i =0; i<N;i++){
            parents[i]= i;
        }
        li.sort(Comparator.comparingInt(x->x[0]));

        long ans = 0;
        int cnt = 0;
        for(int[] edge:li){
            int a = edge[1];
            int b = edge[2];
            if(findParent(parents[a]) != findParent(parents[b])){
                union(a,b);
                ans += edge[0];
                if(++cnt == N)
                    break;
            }
        }
        System.out.println(ans);

    }
    private static int findParent(int n){
        if (parents[n] != n)
            parents[n] = findParent(parents[n]);
        return parents[n];
    }

    private static void union(int a,int b){
        a = findParent(a);
        b = findParent(b);
        if(a<b)
            parents[a] = b;
        else if (a>b)
            parents[b] = a;
    }

}

File: 백준/Gold/16398. 행성 연결/행성 연결.py
N = int(input())
li = [list(map(int, input().split())) for _ in range(N)]
li2 = []

for i in range(N - 1):
    for j in range(i + 1,N):
        li2.append((li[i][j], i, j))

li2.sort()


parent = [i for i in range(N)]
def get_parent(n):
    if parent[n] != n:
        parent[n] = get_parent(parent[n])
    return parent[n]


def union(a, b):
    a = get_parent(a)
    b = get_parent(b)

    if a < b:
        parent[a] = parent[b]
    elif b < a:
        parent[b] = parent[a]
ans = 0
for d, a, b in li2:
    if get_parent(a) != get_parent(b):
        ans += d
        union(a, b)
print(ans)